File: Tests/PUPSISPortalTests/MarkdownCommandsTests.swift
import XCTest
@testable import PUPSISPortal

/// The pure selection transforms behind the notes toolbar.
final class MarkdownCommandsTests: XCTestCase {
    func testWrapSelection() {
        let r = MarkdownCommands.wrap("read chapter", range: NSRange(location: 5, length: 7), with: "**")
        XCTAssertEqual(r.text, "read **chapter**")
        // Inner text stays selected.
        XCTAssertEqual(r.range, NSRange(location: 7, length: 7))
    }

    func testWrapEmptyDropsCursorBetweenMarkers() {
        let r = MarkdownCommands.wrap("", range: NSRange(location: 0, length: 0), with: "*")
        XCTAssertEqual(r.text, "**")
        XCTAssertEqual(r.range, NSRange(location: 1, length: 0))
    }

    func testToggleChecklistAddsThenRemoves() {
        let text = "buy milk\nwash car"
        let all = NSRange(location: 0, length: (text as NSString).length)

        let added = MarkdownCommands.toggleChecklist(text, range: all)
        XCTAssertEqual(added.text, "- [ ] buy milk\n- [ ] wash car")

        let cleared = MarkdownCommands.toggleChecklist(added.text, range: added.range)
        XCTAssertEqual(cleared.text, "buy milk\nwash car")
    }

    func testToggleNumberedIncrements() {
        let text = "one\ntwo\nthree"
        let all = NSRange(location: 0, length: (text as NSString).length)
        let r = MarkdownCommands.toggleNumbered(text, range: all)
        XCTAssertEqual(r.text, "1. one\n2. two\n3. three")
    }

    func testCycleHeadingRoundTrip() {
        var r = MarkdownCommands.cycleHeading("Title", range: NSRange(location: 0, length: 0))
        XCTAssertEqual(r.text, "# Title")
        r = MarkdownCommands.cycleHeading(r.text, range: NSRange(location: 0, length: 0))
        XCTAssertEqual(r.text, "## Title")
        // Past H6 it clears back to plain text.
        r = MarkdownCommands.cycleHeading("###### Title", range: NSRange(location: 0, length: 0))
        XCTAssertEqual(r.text, "Title")
    }

    func testToggleCheckboxFlipsBothWays() {
        XCTAssertEqual(MarkdownCommands.toggleCheckbox("- [ ] task"), "- [x] task")
        XCTAssertEqual(MarkdownCommands.toggleCheckbox("- [x] task"), "- [ ] task")
        XCTAssertEqual(MarkdownCommands.toggleCheckbox("  - [X] indented"), "  - [ ] indented")
    }

    /// Detection underpins the source round-trip: collapsing these to attachments
    /// and reconstructing must reproduce the exact input, so notes never corrupt.
    func testMathDetectionRoundTrips() {
        let text = "Energy $E=mc^2$ and a block:\n$$\\int_0^1 x\\,dx$$\nprice $5 and $10 stays text."
        let matches = MarkdownCommands.mathMatches(in: text)

        // Two math spans; the currency "$5 and $10" is not one of them.
        XCTAssertEqual(matches.count, 2)
        XCTAssertEqual(matches[0].latex, "E=mc^2")
        XCTAssertFalse(matches[0].display)
        XCTAssertEqual(matches[1].latex, "\\int_0^1 x\\,dx")
        XCTAssertTrue(matches[1].display)

        // Reconstruct source from the matches — must equal the original.
        let ns = text as NSString
        var out = ""
        var cursor = 0
        for m in matches {
            out += ns.substring(with: NSRange(location: cursor, length: m.range.location - cursor))
            out += m.display ? "$$\(m.latex)$$" : "$\(m.latex)$"
            cursor = NSMaxRange(m.range)
        }
        out += ns.substring(from: cursor)
        XCTAssertEqual(out, text)
    }

    func testColorWrapsSelection() {
        let r = MarkdownCommands.color("hi", range: NSRange(location: 0, length: 2), hex: "e5484d")
        XCTAssertEqual(r.text, "{#e5484d:hi}")
        XCTAssertEqual(r.range, NSRange(location: 9, length: 2))
    }

    func testColorEmptyDropsCursorInside() {
        let r = MarkdownCommands.color("", range: NSRange(location: 0, length: 0), hex: "2f9e44")
        XCTAssertEqual(r.text, "{#2f9e44:}")
        XCTAssertEqual(r.range, NSRange(location: 9, length: 0))
    }

    func testLinkPlacesCursorOnPlaceholder() {
        let r = MarkdownCommands.link("docs", range: NSRange(location: 0, length: 4))
        XCTAssertEqual(r.text, "[docs](url)")
        XCTAssertEqual(r.range, NSRange(location: 7, length: 3))
    }
}

File: windows/PUPSISPortal.Core.Tests/MarkdownCommandsTests.cs
namespace PUPSISPortal.Core.Tests;

/// <summary>
/// The pure selection transforms behind the notes toolbar.
/// </summary>
public class MarkdownCommandsTests
{
    [Fact]
    public void TestWrapSelection()
    {
        var r = MarkdownCommands.Wrap("read chapter", 5, 7, "**");
        Assert.Equal("read **chapter**", r.Text);
        // Inner text stays selected.
        Assert.Equal(7, r.Start);
        Assert.Equal(7, r.Length);
    }

    [Fact]
    public void TestWrapEmptyDropsCursorBetweenMarkers()
    {
        var r = MarkdownCommands.Wrap("", 0, 0, "*");
        Assert.Equal("**", r.Text);
        Assert.Equal(1, r.Start);
        Assert.Equal(0, r.Length);
    }

    [Fact]
    public void TestToggleChecklistAddsThenRemoves()
    {
        var text = "buy milk\nwash car";
        var all = new { Start = 0, Length = text.Length };

        var added = MarkdownCommands.ToggleChecklist(text, all.Start, all.Length);
        Assert.Equal("- [ ] buy milk\n- [ ] wash car", added.Text);

        var cleared = MarkdownCommands.ToggleChecklist(added.Text, added.Start, added.Length);
        Assert.Equal("buy milk\nwash car", cleared.Text);
    }

    [Fact]
    public void TestToggleNumberedIncrements()
    {
        var text = "one\ntwo\nthree";
        var all = new { Start = 0, Length = text.Length };
        var r = MarkdownCommands.ToggleNumbered(text, all.Start, all.Length);
        Assert.Equal("1. one\n2. two\n3. three", r.Text);
    }

    [Fact]
    public void TestCycleHeadingRoundTrip()
    {
        var r = MarkdownCommands.CycleHeading("Title", 0, 0);
        Assert.Equal("# Title", r.Text);

        r = MarkdownCommands.CycleHeading(r.Text, 0, 0);
        Assert.Equal("## Title", r.Text);

        // Past H6 it clears back to plain text.
        r = MarkdownCommands.CycleHeading("###### Title", 0, 0);
        Assert.Equal("Title", r.Text);
    }

    [Fact]
    public void TestToggleCheckboxFlipsBothWays()
    {
        Assert.Equal("- [x] task", MarkdownCommands.ToggleCheckbox("- [ ] task"));
        Assert.Equal("- [ ] task", MarkdownCommands.ToggleCheckbox("- [x] task"));
        Assert.Equal("  - [ ] indented", MarkdownCommands.ToggleCheckbox("  - [X] indented"));
    }

    /// <summary>
    /// Detection underpins the source round-trip: collapsing these to attachments
    /// and reconstructing must reproduce the exact input, so notes never corrupt.
    /// </summary>
    [Fact]
    public void TestMathDetectionRoundTrips()
    {
        var text = "Energy $E=mc^2$ and a block:\n$$\\int_0^1 x\\,dx$$\nprice $5 and $10 stays text.";
        var matches = MarkdownCommands.MathMatches(text);

        // Two math spans; the currency "$5 and $10" is not one of them.
        Assert.Equal(2, matches.Count);

        Assert.Equal("E=mc^2", matches[0].Latex);
        Assert.False(matches[0].Display);

        Assert.Equal("\\int_0^1 x\\,dx", matches[1].Latex);
        Assert.True(matches[1].Display);

        // Reconstruct source from the matches — must equal the original.
        var out_ = "";
        var cursor = 0;
        foreach (var m in matches)
        {
            out_ += text.Substring(cursor, m.Start - cursor);
            out_ += m.Display ? $"$${m.Latex}$$" : $"${m.Latex}$";
            cursor = m.Start + m.Length;
        }
        out_ += text.Substring(cursor);

        Assert.Equal(text, out_);
    }

    [Fact]
    public void TestColorWrapsSelection()
    {
        var r = MarkdownCommands.Color("hi", 0, 2, "e5484d");
        Assert.Equal("{#e5484d:hi}", r.Text);
        Assert.Equal(9, r.Start);
        Assert.Equal(2, r.Length);
    }

    [Fact]
    public void TestColorEmptyDropsCursorInside()
    {
        var r = MarkdownCommands.Color("", 0, 0, "2f9e44");
        Assert.Equal("{#2f9e44:}", r.Text);
        Assert.Equal(9, r.Start);
        Assert.Equal(0, r.Length);
    }

    [Fact]
    public void TestLinkPlacesCursorOnPlaceholder()
    {
        var r = MarkdownCommands.Link("docs", 0, 4);
        Assert.Equal("[docs](url)", r.Text);
        Assert.Equal(7, r.Start);
        Assert.Equal(3, r.Length);
    }
}
